pilot/cmd/pilot-agent/status/server_test.go

```diff
@@ -312,6 +312,114 @@ my_metric{app="bar"} 0
 	}
 }
 
+func TestStatsContentType(t *testing.T) {
+	appOpenMetrics := `# TYPE jvm info
+# HELP jvm VM version info
+jvm_info{runtime="OpenJDK Runtime Environment",vendor="AdoptOpenJDK",version="16.0.1+9"} 1.0
+# TYPE jmx_config_reload_success counter
+# HELP jmx_config_reload_success Number of times configuration have successfully been reloaded.
+jmx_config_reload_success_total 0.0
+jmx_config_reload_success_created 1.623984612719E9
+# EOF
+`
+	appText004 := `# HELP jvm_info VM version info
+# TYPE jvm_info gauge
+jvm_info{runtime="OpenJDK Runtime Environment",vendor="AdoptOpenJDK",version="16.0.1+9",} 1.0
+# HELP jmx_config_reload_failure_created Number of times configuration have failed to be reloaded.
+# TYPE jmx_config_reload_failure_created gauge
+jmx_config_reload_failure_created 1.624025983489E9
+`
+	envoy := `# TYPE my_metric counter
+my_metric{} 0
+# TYPE my_other_metric counter
+my_other_metric{} 0
+`
+	output := `# TYPE my_metric counter
+my_metric{} 0
+# TYPE my_other_metric counter
+my_other_metric{} 0
+# HELP jvm_info VM version info
+# TYPE jvm_info gauge
+jvm_info{runtime="OpenJDK Runtime Environment",vendor="AdoptOpenJDK",version="16.0.1+9",} 1.0
+`
+
+	cases := []struct {
+		name             string
+		acceptHeader     string
+		expectParseError bool
+	}{
+		{
+			name:             "openmetrics accept header",
+			acceptHeader:     `application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1`,
+			expectParseError: false,
+		},
+		{
+			name:             "plaintext accept header",
+			acceptHeader:     (string)(expfmt.FmtText),
+			expectParseError: false,
+		},
+		{
+			name:             "empty accept header",
+			acceptHeader:     "",
+			expectParseError: false,
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			rec := httptest.NewRecorder()
+			envoy := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if _, err := w.Write([]byte(envoy)); err != nil {
+					t.Fatalf("write failed: %v", err)
+				}
+			}))
+			defer envoy.Close()
+			app := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				format := expfmt.NegotiateIncludingOpenMetrics(r.Header)
+				var negotiatedMetrics string
+				if format == expfmt.FmtOpenMetrics {
+					negotiatedMetrics = appOpenMetrics
+				} else if format == expfmt.FmtText {
```

Comment:
Are metrics in FmtText not compatible with FmtOpenMetrics? My (very loose) understanding was FmtOpenMetrics is just adding some additional legal syntax, so its more of a superset of FmtText. Is that not accurate?

Reply:
I also am not an expert, but there seems to be differences in the actual metric types in `FmtOpenMetrics` that are not present in `FmtText`. Notably, `FmtOpenMetrics` has support for [the `info` type](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#info) which isn't a valid type in `FmtText`.

Reply:
Right, what I was thinking is:

* Prometheus sets Accept=FmtText. All apps/envoy/agent must return FmtText, and everything works
* Prometheus sets Accept=FmtOpenMetrics. The app returns the new FmtOpenMetrics, but envoy returns FmtText since that is all it supports. Because FmtText is a subset of FmtOpenMetrics, prometheus interprets it all correctly

We should never have a case where prometheus sets Accept=FmtText and gets back FmtOpenMetrics(?)


Is the issue when we respond we don't properly tell Prometheus which format we actually responded with (FmtOpenMetrics)? And we should add some response header?

Reply:
Got it, I see what you are getting at. 

In theory we should be able to:

- pass through the `Accept` from Prom to the application container, i.e. app will correctly negotiate `Content-Type` from Prom's `Accept`
- based on `Content-Type` from app container negotiation, set response header for merged metrics to this `Content-Type` (either FmtText or FmtOpenMetrics). Since agent and envoy metrics will be FmtText, they should be compatible with either format
- write out app container metrics last to support [the "# EOF" termination sequence required by FmtOpenMetrics](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#overall-structure)

A quick test with this approach exposes another issue; it seems Prometheus' FmtOpenMetrics parser does not like blank lines in the metrics, such as those in the envoy metrics, e.g.:
```
# TYPE envoy_cluster_assignment_stale counter
envoy_cluster_assignment_stale{cluster_name="xds-grpc"} 0

# TYPE envoy_cluster_assignment_timeout_received counter
envoy_cluster_assignment_timeout_received{cluster_name="xds-grpc"} 0

# TYPE envoy_cluster_bind_errors counter
envoy_cluster_bind_errors{cluster_name="xds-grpc"} 0
```

The end result is a parse error: `"INVALID" "\n" is not a valid start token`.

I also further tested this out by adding a blank line in Prom's parse testing: https://github.com/prometheus/prometheus/blob/main/pkg/textparse/openmetricsparse_test.go#L32-L33, which fails with the same parse error:
```
--- FAIL: TestOpenMetricsParse (0.00s)
    /Users/law/codebase/prometheus/pkg/textparse/openmetricsparse_test.go:225: 
        	Error Trace:	openmetricsparse_test.go:225
        	Error:      	Received unexpected error:
        	            	"INVALID" "\n" is not a valid start token
        	            	github.com/prometheus/prometheus/pkg/textparse.(*OpenMetricsParser).Next
        	            		/Users/law/codebase/prometheus/pkg/textparse/openmetricsparse.go:355
        	            	github.com/prometheus/prometheus/pkg/textparse.TestOpenMetricsParse
        	            		/Users/law/codebase/prometheus/pkg/textparse/openmetricsparse_test.go:221
        	            	testing.tRunner
        	            		/usr/local/Cellar/go/1.16/libexec/src/testing/testing.go:1194
        	            	runtime.goexit
        	            		/usr/local/Cellar/go/1.16/libexec/src/runtime/asm_amd64.s:1371
        	Test:       	TestOpenMetricsParse
FAIL
FAIL	github.com/prometheus/prometheus/pkg/textparse	0.126s
FAIL
```

Lastly, a quick hacked together build with the envoy metrics removed was scraped successfully from Prometheus.
In that case, I had istio agent metrics written as FmtText then app metrics written as FmtOpenMetrics and a response `Content-Type` of FmtOpenMetrics.

So unfortunately I don't think the existing envoy metrics are compatible with FmtOpenMetrics.

Reply:
> So unfortunately I don't think the existing envoy metrics are compatible with FmtOpenMetrics.

Hmm.. I wonder if this is something we can change in Envoy?

The reason I am concerned with the PR as-is is it means applications cannot use OpenMetrics

@douglas-reid @ramaraochavali thoughts?

Reply:
Sure enough, the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#abnf) doesn't include a provision for empty lines.

Instead of forcing only one format, can we make the merging do the conversion?  Surely we can parse and translate in-agent, right?

Reply:
> Hmm.. I wonder if this is something we can change in Envoy?

I believe so, I'll open an issue for this.

>Instead of forcing only one format, can we make the merging do the conversion?

This seems viable, I'll test it out